Modules/Scripted/VectorToScalarVolume/VectorToScalarVolume.py

```diff
@@ -26,82 +63,420 @@ def __init__(self, parent):
+                               self.conversionMethodWidget.conversionMethod(),
+                               self.conversionMethodWidget.componentToExtract())
+
+    # make the output volume appear in all the slice views
+    if success:
+      selectionNode = slicer.app.applicationLogic().GetSelectionNode()
+      selectionNode.SetReferenceActiveVolumeID(self.outputVolumeNode().GetID())
+      slicer.app.applicationLogic().PropagateVolumeSelection(0)
+
+
+#
+# VectorToScalarVolumeConversionMethodWidget
+#
+
+class VectorToScalarVolumeConversionMethodWidget(qt.QWidget):
+  """
+  Widget to interact with conversion parameters.
+  """
+
+  EXTRACT_COMPONENT_NONE = -1
+
+  def __init__(self, parent=None):
+    qt.QWidget.__init__(self, parent)
+    self._inputVolumeNode = None
+    self._conversionMethod = None
+    self._componentToExtract = self.EXTRACT_COMPONENT_NONE
+    self.setup()
+
+  def setup(self):
+    self.methodLayout = qt.QHBoxLayout(self)
+    self.methodSelectorComboBox = qt.QComboBox()
+    self.methodSelectorComboBox.insertItem(1, "Single Component Extraction", VectorToScalarVolumeLogic.SINGLE_COMPONENT)
+    self.methodSelectorComboBox.setItemData(1, 'Extract single component', qt.Qt.ToolTipRole)
+    self.methodSelectorComboBox.insertItem(2, "Luminance", VectorToScalarVolumeLogic.LUMINANCE)
+    self.methodSelectorComboBox.setItemData(2, '(RGB,RGBA) Luminance from first three components: 0.30*R + 0.59*G + 0.11*B + 0.0*A)', qt.Qt.ToolTipRole)
+    # XXX: Average could be extended to let the user choose the weights of the components.
+    self.methodSelectorComboBox.insertItem(3, "Average", VectorToScalarVolumeLogic.AVERAGE)
+    self.methodSelectorComboBox.setItemData(3, 'Average all the components.', qt.Qt.ToolTipRole)
+
+    self.methodLayout.addWidget(self.methodSelectorComboBox)
+    self.methodLayout.addItem(qt.QSpacerItem(40, 20, qt.QSizePolicy.Expanding, qt.QSizePolicy.Minimum))
+
+    # Option: Single component
+    singleComponentLayout = qt.QHBoxLayout()
+    self.componentsComboBox = qt.QComboBox()
+    singleComponentLayout.addWidget(self.componentsComboBox)
+    singleComponentLayout.addItem(qt.QSpacerItem(40, 20, qt.QSizePolicy.Expanding, qt.QSizePolicy.Minimum))
+    self.methodLayout.addLayout(singleComponentLayout)
+
+    # Connections
+    # setConversionMethod uses Qt::UserRole data as input, not index. The connection makes the conversion from the index to data.
+    self.methodSelectorComboBox.connect('currentIndexChanged(int)',
+                                        lambda currentIndex:
+                                        self.setConversionMethod(self.methodSelectorComboBox.itemData(currentIndex))
+                                        )
+
+    self.componentsComboBox.connect('currentIndexChanged(int)', self.setComponentToExtract)
+
+  def inputVolumeNode(self):
+    return self._inputVolumeNode
+
+  def setInputVolumeNode(self, node):
+    if isinstance(node, basestring):
```

Comment:
This kind of ambiguity - accepting a node or some string (node id? node name? what if there are multiple nodes with the same name?) - is acceptable in testing and development tools for convenience, but not in production code.

For example, slicer.util.getNode has this behavior exactly because it is often used in the Python console to interactively get a node. However, in modules you would use slicer.mrmlScene.GetNode()or slicer.mrmlScene.GetFirstNodeByName().

Since the widget is not supposed to be used from any other modules (you only interact through MRML nodes and sometimes by using another module's logic), I would remove this method completely.

Reply:
The `getNode` function only lookup node ID (defined at the top of the file). There are no risk of finding a node by name.

The issue is that the name of the function is the same as `slicer.util.getNode`

Reply:
OK then. Still, I find all these set/get functions unnecessary, as the updateMRML… and updateGUI… methods take care of keeping the widget and MRML node in sync.

Reply:
Removed variables in widget, the values are only stored in the ComboBoxes. And the 'upper' widget makes the connections to keep updated the parameterNode.